x-pack/plugins/apm/server/lib/service_map/get_service_map_from_trace_ids.test.ts

```diff
@@ -5,24 +5,198 @@
+      });
+
+      const connectionsPairs = getConnectionsPairs(connections);
+
+      expect(connectionsPairs).toEqual([
+        'opbeans-python:testing -> opbeans-node:testing',
+      ]);
+    });
+  });
+
+  describe('if service name is given', () => {
+    it('shows all connections for opbeans-node', () => {
+      const connections = getConnections({
+        paths,
+        serviceName: 'opbeans-node',
+        environment: undefined,
+      });
+
+      const connectionsPairs = getConnectionsPairs(connections);
+
+      expect(connectionsPairs).toEqual([
+        'opbeans-python:testing -> opbeans-node:testing',
+      ]);
+    });
+  });
+
+  describe('if environment is given', () => {
+    it('shows all connections for testing environment', () => {
+      const connections = getConnections({
+        paths,
+        serviceName: undefined,
+        environment: 'testing',
+      });
+
+      const connectionsPairs = getConnectionsPairs(connections);
+
+      expect(connectionsPairs).toEqual([
+        'opbeans-python:testing -> opbeans-node:testing',
+      ]);
+    });
+    it('shows all connections for production environment', () => {
+      const connections = getConnections({
+        paths,
+        serviceName: undefined,
+        environment: 'production',
+      });
+
+      const connectionsPairs = getConnectionsPairs(connections);
+
+      expect(connectionsPairs).toEqual([
+        'opbeans-go:production -> opbeans-java:production',
+      ]);
+    });
+  });
+
+  describe('if environment is "not defined"', () => {
+    it('shows all connections where environment is not set', () => {
+      const environmentNotDefinedPaths = [
+        [
+          {
+            'service.environment': null,
+            'service.name': 'opbeans-go',
+            'agent.name': 'go',
+          },
+          {
+            'service.environment': null,
+            'service.name': 'opbeans-java',
+            'agent.name': 'java',
+          },
+          {
+            'span.subtype': 'http',
+            'span.destination.service.resource': '172.18.0.6:3000',
+            'span.type': 'external',
+          },
+        ],
+        [
+          {
+            'service.environment': null,
+            'service.name': 'opbeans-python',
+            'agent.name': 'python',
+          },
+          {
+            'service.environment': null,
+            'service.name': 'opbeans-node',
+            'agent.name': 'nodejs',
+          },
+          {
+            'span.subtype': 'http',
+            'span.destination.service.resource': '172.18.0.6:3000',
+            'span.type': 'external',
+          },
+        ],
+      ] as ConnectionNode[][];
+      const connections = getConnections({
+        paths: environmentNotDefinedPaths,
+        serviceName: undefined,
+        environment: 'ENVIRONMENT_NOT_DEFINED',
+      });
+
+      const connectionsPairs = getConnectionsPairs(connections);
+      expect(connectionsPairs).toEqual([
```

Comment:
suppose we have a bug where filtering on ENVIRONMENT_NOT_DEFINED returns all paths/connections. Would this test catch it?

Reply:
no, because we are mocking the data to have `service.environment: null`. If I filter out external node (`span.type !== 'external'`) as you suggested [here](https://github.com/elastic/kibana/pull/77757#discussion_r496576586), then I can create a test for this case.

Reply:
I'm not sure I understand, can you clarify?

Reply:
> suppose we have a bug where filtering on ENVIRONMENT_NOT_DEFINED returns all paths/connections

if it happens this wouldn't be the result of paths, right?
```js
const environmentNotDefinedPaths = [
        [
          {
            'service.environment': null,
            'service.name': 'opbeans-go',
            'agent.name': 'go',
          },
          {
            'service.environment': null,
            'service.name': 'opbeans-java',
            'agent.name': 'java',
          },
          {
            'span.subtype': 'http',
            'span.destination.service.resource': '172.18.0.6:3000',
            'span.type': 'external',
          },
        ],
        [
          {
            'service.environment': null,
            'service.name': 'opbeans-python',
            'agent.name': 'python',
          },
          {
            'service.environment': null,
            'service.name': 'opbeans-node',
            'agent.name': 'nodejs',
          },
          {
            'span.subtype': 'http',
            'span.destination.service.resource': '172.18.0.6:3000',
            'span.type': 'external',
          },
        ],
      ] as ConnectionNode[][];
```

Reply:
Still not sure I understand 😅 Because there are no environments in `environmentNotDefinedPaths`, and you're only using those paths in `shows all connections where environment is not set`, this test doesn't test whether paths that do & only have values for service.environment are excluded. LMK if you want to Zoom about it.

Reply:
😅 to answer this question:
>suppose we have a bug where filtering on ENVIRONMENT_NOT_DEFINED returns all paths/connections. Would this test catch it?

No, this test won't catch if there is a bug  filtering on ENVIRONMENT_NOT_DEFINED.